recipes/newrelic/apm/php-agent/php.yml

```diff
@@ -110,12 +190,12 @@ install:
           # Web server was detected. Send transaction associated with the webserver php.ini
           # and the newrelic.ini file associated with that web server.
           #
-             php -c "${WEB_PHP_INI_DIR}/php.ini" -c "${WEB_NR_INI_DIR}/newrelic.ini" -n --ini &>/dev/null
-             php -c "${WEB_PHP_INI_DIR}/php.ini" -c "${WEB_NR_INI_DIR}/newrelic.ini" -n --ini &>/dev/null
-             php -c "${WEB_PHP_INI_DIR}/php.ini" -c "${WEB_NR_INI_DIR}/newrelic.ini" -n --ini &>/dev/null
+             sudo -u ${nonpriv_user} php -c "${WEB_PHP_INI_DIR}/php.ini" -c "${WEB_NR_INI_DIR}/newrelic.ini" -n --ini &>/dev/null
+             sudo -u ${nonpriv_user} php -c "${WEB_PHP_INI_DIR}/php.ini" -c "${WEB_NR_INI_DIR}/newrelic.ini" -n --ini &>/dev/null
+             sudo -u ${nonpriv_user} php -c "${WEB_PHP_INI_DIR}/php.ini" -c "${WEB_NR_INI_DIR}/newrelic.ini" -n --ini &>/dev/null
```

Comment:
In the FPM + Apache scenario where Apache isn't configured to use FPM, this may run these processes as the wrong user. Restarting both processes, as a way of avoiding the need to detect if FPM is really in use, fails in this scenario as we have to choose a user to use here which could introduce permissions problems if we choose incorrectly. Switching to a URL based approach would avoid this complexity at the cost of another user prompt.

Reply:
Few things about this:
When the processes are restarted, they restart the daemon so it is already running when this call is made so the non-privileged user used here does not affect the daemon.

I considered looping through the non-privileged users and doing the php call that way; however, the `newrelic-install.sh` script only installs `newrelic.ini` in one directory anyway.  If both fpm and apache exist on the system, the install script will only put the `newrelic.ini` file in one location.  So this is calling only the once with the priority non-privileged user because that is literally the only non-cli place our install will put the newrelic.ini file.

URL-based adds complexity and decreases reliability.  URL based approach is not just complexity of another user prompt (although everytime a user inputs data is a chance for wrong data to be entered) and the complexity of trying to guess what it should be in case of silent mode (which complexity increases once we move to dockerized support).  It is also the complexity of triggering a PHP page load.  What if they url they give/we guess doesn't have php to trigger the agent application?  Wordpress is notorious for not loading plugins with curl (they have their own functions to test).  I experimented a lot with this could not get the wordpress site to generate php traffic ( in order to trigger the initializing transaction )when using curl.  So to make it work reliably, we'd have to generate a test page (like one with phpinfo) and put it as a best guess in their webserver assuming we can locate it and have read access to it.

Reply:
If the process gets restarted as a privileged user and this uses a non-privileged user, don't we run the risk that the PHP cli process will be unable to communicate with the daemon?

Reply:
The started PHP process needs to be able to access the created socket. If the socket was created by the PHP process running as root, it might not be readable by a PHP process running as ubuntu (depending on the umask setting). I thought this was the core of the permissions issues caused by running PHP processes as different users.

Reply:
The started PHP process needs to be able to access the created socket. If the socket was created by the PHP process running as root, it might not be readable by a PHP process running as ubuntu (depending on the umask setting). I thought this was the core of the permissions issues caused by running PHP processes as different users.

Reply:
I'm experimenting with this a bit, expect to have more details tomorrow.

Reply:
Permissions issues are not related to anything different we should do with the virtuoso install as they are more systemic and should be fixed at a higher level.

See related [issue](https://github.com/newrelic/newrelic-php-agent/issues/168) for details.